tests/test_environment.py

```diff
@@ -1,4 +1,7 @@
+from pathlib import Path
+
 from helpers import run_pipx_cli
+from pipx.constants import load_dir_from_environ
```

Comment:
Probably call the `environment` command instead of importing (to comply with this test files' remaining contents)?

Reply:
I tried by mocking the environment variable at the start of the test. While that made it show up under `Environment variables (set by user):`, it did now show in `Derived values (computed by pipx):`, because the globals were not evaluated again, even when calling `importlib.refresh` on `constants`. 

Got a suggestion on how to fix this?

Reply:
> Got a suggestion on how to fix this?

I'll have a look.

Reply:
I've had another look at this and I think I could only make the `environment` command work in this test by:
- Reworking how the constants work, e.g. by storing them in an object instead (I hope someone will do this when/if https://github.com/pypa/pipx/pull/1088#issuecomment-1934516696 is picked up again)
- Have some code in the constants that allows recomputing the constant in question which I can call from the test suite. This would duplicate the creation logic though and add complexity to the constants.
```python
LOCAL_BIN_DIR = load_dir_from_environ("PIPX_BIN_DIR", DEFAULT_PIPX_BIN_DIR)

def recompute_bin_dir():
    global LOCAL_BIN_DIR
    LOCAL_BIN_DIR = load_dir_from_environ("PIPX_BIN_DIR", DEFAULT_PIPX_BIN_DIR)
```

IMO while this test is not quite in line with the existing tests is `test_environments`, it does a good job of testing the logic in question and it's not worth making the constants more complicated.

Reply:
> While that made it show up under `Environment variables (set by user):`, it did now show in `Derived values (computed by pipx):`

Isn't this expected behaviour, BTW? Or why do you think this isn't sufficient to verify the behaviour in question? I won't block this, though.

Reply:
No, the derived value should be based on the environment variable, but it isn't when running in the tests because the constants are not reevaluated.

This is the output if I run it normally (See `PIPX_BIN_DIR` in environment variables and `PIPX_BIN_DIR` in derived values):
```
(develop-3-12) ➜  pipx git:(resolve_home_path_in_dir_env_vars) ✗ PIPX_BIN_DIR="~/test" pipx environment
Environment variables (set by user):

PIPX_HOME=
PIPX_BIN_DIR=~/test
PIPX_MAN_DIR=
PIPX_SHARED_LIBS=
PIPX_DEFAULT_PYTHON=
USE_EMOJI=

Derived values (computed by pipx):

PIPX_HOME=/home/robert/.local/pipx
PIPX_BIN_DIR=/home/robert/test
PIPX_MAN_DIR=/home/robert/.local/share/man
PIPX_SHARED_LIBS=/home/robert/.local/pipx/shared
PIPX_LOCAL_VENVS=/home/robert/.local/pipx/venvs
PIPX_LOG_DIR=/home/robert/.local/pipx/logs
PIPX_TRASH_DIR=/home/robert/.local/pipx/.trash
PIPX_VENV_CACHEDIR=/home/robert/.local/pipx/.cache
PIPX_DEFAULT_PYTHON=/home/robert/private/code/pipx/.nox/develop-3-12/bin/python
USE_EMOJI=true
```

And this is the result if I patch the `PIPX_BIN_DIR` env variable in a test and run `assert not run_pipx_cli(["environment"])` (See `PIPX_BIN_DIR` being set correctly in the environment variables, but not in the derived values):

```
Environment variables (set by user):

PIPX_HOME=
PIPX_BIN_DIR=~/test
PIPX_MAN_DIR=
PIPX_SHARED_LIBS=
PIPX_DEFAULT_PYTHON=
USE_EMOJI=

Derived values (computed by pipx):

PIPX_HOME=/home/robert/.local/pipx
PIPX_BIN_DIR=/home/robert/.local/bin
PIPX_MAN_DIR=/home/robert/.local/share/man
PIPX_SHARED_LIBS=/home/robert/.local/pipx/shared
PIPX_LOCAL_VENVS=/home/robert/.local/pipx/venvs
PIPX_LOG_DIR=/home/robert/.local/pipx/logs
PIPX_TRASH_DIR=/home/robert/.local/pipx/.trash
PIPX_VENV_CACHEDIR=/home/robert/.local/pipx/.cache
PIPX_DEFAULT_PYTHON=/home/robert/private/code/pipx/.nox/tests-3-12/bin/python
USE_EMOJI=true
```

Reply:
Yes, obviously monkeypatching takes place after the constants have been initialised. I agree there is nothing to do here, right now.